webrender/src/device/gfx.rs

```diff
@@ -708,75 +700,68 @@ pub struct Buffer<B: hal::Backend> {
 impl<B: hal::Backend> Buffer<B> {
     pub fn new(
         device: &B::Device,
-        memory_types: &[hal::MemoryType],
-        usage: hal::buffer::Usage,
-        pitch_alignment_mask: usize,
+        heaps: &mut Heaps<B>,
+        memory_usage: MemoryUsageValue,
+        buffer_usage: hal::buffer::Usage,
+        alignment_mask: usize,
         data_len: usize,
         stride: usize,
     ) -> Self {
-        let buffer_size = (data_len * stride + pitch_alignment_mask) & !pitch_alignment_mask;
+        let buffer_size = (data_len * stride + alignment_mask) & !alignment_mask;
         let mut buffer = unsafe {
             device
-                .create_buffer(buffer_size as u64, usage)
+                .create_buffer(buffer_size as u64, buffer_usage)
                 .expect("create_buffer failed")
         };
         let requirements = unsafe { device.get_buffer_requirements(&buffer) };
-        let memory_type = memory_types
-            .iter()
-            .enumerate()
-            .position(|(id, mem_type)| {
-                requirements.type_mask & (1 << id) != 0
-                    && mem_type
-                        .properties
-                        .contains(hal::memory::Properties::CPU_VISIBLE)
-            })
-            .unwrap()
-            .into();
-        let memory = unsafe { device.allocate_memory(memory_type, requirements.size) }
-            .expect("Allocate memory failed");
-        unsafe { device.bind_buffer_memory(&memory, 0, &mut buffer) }
+
+        let memory_block = heaps.allocate(
+            device,
+            requirements.type_mask as u32,
+            memory_usage,
+            requirements.size,
+            std::cmp::max(requirements.alignment, (alignment_mask + 1) as u64)
+        ).expect("Allocate memory failed");
+
+        unsafe { device.bind_buffer_memory(&memory_block.memory(), memory_block.range().start, &mut buffer) }
             .expect("Bind buffer memory failed");
 
-        let buffer = Buffer {
-            memory,
+        Buffer {
+            memory_block,
             buffer,
             buffer_size: requirements.size as _,
             buffer_len: data_len,
             stride,
             state: Cell::new(hal::buffer::Access::empty()),
-        };
-        buffer
+        }
     }
 
-    pub fn update_all<T: Copy>(&self, device: &B::Device, data: &[T]) {
-        let mut upload_data = unsafe {
-            device.acquire_mapping_writer::<T>(&self.memory, 0 .. self.buffer_size as u64)
+    pub fn update_all<T: Copy>(&mut self, device: &B::Device, data: &[T]) {
+        let offset = self.memory_block.range().start;
+        let length = (data.len() * std::mem::size_of::<T>()) as u64;
+        let range = offset .. offset + length;
+        unsafe {
+            let mut mapped = self.memory_block.map(device, range).expect("Mapping memory block failed");
```

Comment:
something is not right here. The memory block gets mapped, and the offset we provide already includes the memory block's start offset? Sounds like it is applied twice

Reply:
Yes the offset is applied twice, but without that we don't draw properly. I have checked our side, but didn't found any clue. I will look into rendy, what is happening with these ranges.

Reply:
Rendy should treat range specified in `Block::map` as subrange of the block's range.
It will add `Block::range().start` to both sides and ensure it doesn't go further than `Block::range().end`

Reply:
Could you try to debug log values of blocks and ranges to see where it fails to calculate correct offsets?

Reply:
Yes, I have started to look into that. But found another issue: we only use dedicated allocator for allocation. Maybe my configuration isn't correct. After I figured out why this happens,I will continue debugging the ranges.

Reply:
If this block was allocated using dedicated allocator than `offset` must always be `0`